src/runtime/virtcontainers/nydus.go

```diff
@@ -0,0 +1,364 @@
+	extraArgs   []string
+}
+
+func (nd *nydusd) Start(ctx context.Context, onQuit onQuitFunc) (int, error) {
+	span, _ := katatrace.Trace(ctx, nd.Logger(), "Start", nydusdTracingTags)
+	defer span.End()
+	pid := 0
+
+	args := []string{
+		"--log-level", nd.logLevel,
+		"--apisock", nd.apiSockPath,
+		"--sock", nd.sockPath,
+	}
+	if len(nd.extraArgs) > 0 {
+		args = append(args, nd.extraArgs...)
+	}
+	cmd := exec.Command(nd.path, args...)
+	r, w, err := os.Pipe()
+	if err != nil {
+		return pid, err
+	}
+	cmd.Stdout = w
+	cmd.Stderr = w
+	fields := logrus.Fields{
+		"path": nd.path,
+		"args": strings.Join(args, " "),
+	}
+	nd.Logger().WithFields(fields).Info()
+
+	// Create nydusd in shim netns as it needs to access host network
+	if err = doNetNS(shimNsPath, func(_ ns.NetNS) error {
+		return cmd.Start()
+	}); err != nil {
+		return pid, err
+	}
+
+	// Monitor nydusd's stdout/stderr and stop sandbox if nydusd quits
+	go func() {
+		scanner := bufio.NewScanner(r)
+		for scanner.Scan() {
+			nd.Logger().Info(scanner.Text())
+		}
+		nd.Logger().Info("nydusd quits")
+		// Wait to release resources of nydusd process
+		cmd.Process.Wait()
+		if onQuit != nil {
+			onQuit()
+		}
+	}()
+	if err := nd.setupPassthroughFS(); err != nil {
+		return pid, err
+	}
+	nd.pid = cmd.Process.Pid
+	return nd.pid, nil
+}
+
+func (nd *nydusd) setupPassthroughFS() error {
+	if err := os.MkdirAll(nd.sourcePath, DirMode); err != nil {
+		return err
+	}
+	nc, err := NewNydusClient(nd.apiSockPath)
+	if err != nil {
+		return err
+	}
+	nd.Logger().WithField("from", nd.sourcePath).
+		WithField("dest", sharedPathInGuest).Info("prepare mount passthroughfs")
+
+	mr := NewMountRequest(nydusPassthroughfsType, nd.sourcePath, "")
+	return nc.Mount(sharedPathInGuest, mr)
+}
+
+func (nd *nydusd) Mount(opt MountOption) error {
+	nc, err := NewNydusClient(nd.apiSockPath)
+	if err != nil {
+		return err
+	}
+	nd.Logger().WithField("from", opt.source).
+		WithField("dest", opt.mountpoint).Info("prepare mount rafs")
+
+	mr := NewMountRequest(nydusRafsType, opt.source, opt.config)
+	return nc.Mount(opt.mountpoint, mr)
+}
+
+func (nd *nydusd) Umount(mountpoint string) error {
+	nc, err := NewNydusClient(nd.apiSockPath)
+	if err != nil {
+		return err
+	}
+	nd.Logger().WithField("mountpoint", mountpoint).Info("umount rafs")
+	return nc.Umount(mountpoint)
+}
+
+func (nd *nydusd) Stop(ctx context.Context) error {
+	if err := nd.kill(ctx); err != nil {
+		nd.Logger().WithError(err).WithField("pid", nd.pid).Warn("kill nydusd failed")
+		return nil
+	}
+
+	err := os.Remove(nd.sockPath)
+	if err != nil {
+		nd.Logger().WithError(err).WithField("path", nd.sockPath).Warn("removing nydusd socket failed")
+	}
+	return nil
+}
+
+func (nd *nydusd) Logger() *logrus.Entry {
+	return virtLog.WithField("subsystem", "nydusd")
+}
+
+func (nd *nydusd) kill(ctx context.Context) (err error) {
+	span, _ := katatrace.Trace(ctx, nd.Logger(), "kill", nydusdTracingTags)
+	defer span.End()
+
+	if nd.pid == 0 {
```

Comment:
`nd.pid` is an `int` but there is no check to see whether it is negative. If it is allowed (to kill the process group) that's fine, but otherwise there should be a check in this function (and ideally a unit test).

Reply:
yeah, have updated to ni.pid <=0, maybe add a unit test later